File: prolog/sos/sos2.pl
% Second-Order System Λ∀,#
:- op(600,xfx,[::,#]).
:- op(650,yfx,[$,!]).
:- op(920,xfx,[⟹,▷,⊢,⟹*]).
:- op(1200,xfx,[--]).
:- use_module(rtg).
term_expansion(A--B,R) :- next_term_expansion(B:-A,R).
% Syntaxs

syntax(k).
syntax(t). t(T) :- atom(T), \+b(T). 
b ::= int | bool.
l ::= atom.
list(A) ::= [] | [A | list(A)].
record(A) ::= list(A).
variant(A) ::= {list(A)}.
σ ::= b | t | (σ->σ) | record(l:σ) | variant(l:σ) | ∀(t,k,σ).
k ::= u | record(l:σ) | variant(l:σ).
syntax(integer).
syntax(x). x(X) :- atom(X), \+cb(X).
i ::= integer.
cb ::= true | false | i.
'M' ::= x | cb | λ(x:σ,'M') | ('M' $ 'M') | λ(t::k,'M') | ('M' ! σ)
    | record(l='M') | 'M'#l | modify('M',l,'M')
    | {[l='M']}:σ | case('M',variant(l='M')).
q(Q) :- σ(Q).
m(M) :- 'M'(M).
:- begin_var_names(['^[σlxktcbi]'],['^(true|bool|int)$']).

% Kinding rules

K ⊢     t : lσs  :- member(t:lσ2s,K), intersection(lσ2s,lσs,lσs).
_ ⊢  lσ2s : lσs  :- intersection(lσ2s,lσs,lσs).
K ⊢     t :{lσs} :- member(t:{lσ2s},K), intersection(lσ2s,lσs,lσs).
_ ⊢ {lσ2s}:{lσs} :- intersection(lσ2s,lσs,lσs).
_ ⊢     _ : u.

% Substitutions

tsub(S,t,N_) :- t(t),member(N/t,S),!,tsub(S,N,N_).
tsub(_,t,t) :- t(t),!.
tsub(_,b,b) :- b(b),!.
tsub(S,(σ1->σ2),(σ1_->σ2_)) :- tsub(S,σ1,σ1_),tsub(S,σ2,σ2_).
tsub(S,lMs,lMs_) :- maplist({S}/[l:M,l:M_]>>tsub(S,M,M_),lMs,lMs_).
tsub(S,{lMs},{lMs_}) :- maplist({S}/[l:M,l:M_]>>tsub(S,M,M_),lMs,lMs_).
tsub(S,∀(t,k,σ),∀(t,k_,σ_)) :- subtract(S,[_/t],S_),ksub(S_,k,k_),tsub(S_,σ,σ_).

ksub(_,u,u).
ksub(S,lσs,lσs_) :- maplist({S}/[l:σ,l:σ_]>>tsub(S,σ,σ_), lσs,lσs_).
ksub(S,{lσs},{lσs_}) :- maplist({S}/[l:σ,l:σ_]>>tsub(S,σ,σ_), lσs,lσs_).

msub(S,x,N_) :- x(x),member(N/x,S),!,msub(S,N,N_).
msub(_,x,x) :- x(x),!.
msub(_,cb,cb) :- cb(cb),!.
msub(S,λ(x:σ,M),λ(x:σ,M_)) :- subtract(S,[_/x],S_),!,msub(S_,M,M_).
msub(S,(M1$M2),(M1_$M2_)) :- msub(S,M1,M1_), msub(S,M2,M2_).
msub(S,(M!σ),(M_!σ)) :- msub(S,M,M_).
msub(S,λ(t::k,M),λ(t::k,M_)) :- msub(S,M,M_).
msub(S,lMs,lMs_) :- maplist({S}/[l=M,l=M_]>>msub(S,M,M_),lMs,lMs_).
msub(S,(M#l),(M_#l)) :- msub(S,M,M_).
msub(S,modify(M1,l,M2),modify(M1_,l,M2_)) :- msub(S,M1,M1_), msub(S,M2,M2_).
msub(S,{[l=M]}:σ,{[l=M_]}:σ) :- msub(S,M,M_).
msub(S,case(M,{lMs}),case(M_,{lMs_})) :- msub(S,M,M_),maplist({S}/[l=Mi,l=Mi_]>>msub(S,Mi,Mi_),lMs,lMs_).

mtsub(S,λ(x:σ,M),λ(x:σ_,M_)) :- tsub(S,σ,σ_),mtsub(S,M,M_).
mtsub(S,(M1$M2),(M1_$M2_)) :- mtsub(S,M1,M1_), mtsub(S,M2,M2_).
mtsub(S,(M!σ),(M_!σ_)) :- mtsub(S,M,M_), tsub(S,σ,σ_).
mtsub(S,λ(t::k,M),λ(t_::k_,M_)) :- subtract(S,[_/t],S_),tsub(S_,t,t_),ksub(S_,k,k_),mtsub(S_,M,M_).
mtsub(S,lMs,lMs_) :- maplist({S}/[l=M,l=M_]>>mtsub(S,M,M_),lMs,lMs_).
mtsub(S,(M#l),(M_#l)) :- mtsub(S,M,M_).
mtsub(S,modify(M1,l,M2),modify(M1_,l,M2_)) :- mtsub(S,M1,M1_), mtsub(S,M2,M2_).
mtsub(S,{[l=M]}:σ,{[l=M_]}:σ) :- mtsub(S,M,M_).
mtsub(S,case(M,{lMs}),case(M_,{lMs_})) :- mtsub(S,M,M_),maplist({S}/[l=Mi,l=Mi_]>>mtsub(S,Mi,Mi_),lMs,lMs_).
mtsub(_,M,M).

% Reduction rules

(λ(x:_,M)$N)             ⟹ M_              :- msub([N/x], M,M_).       % (β)
(λ(t::_,M)!σ)            ⟹ M_              :- mtsub([σ/t], M,M_).      % (type-β)
lMs#li                   ⟹ Mi              :- member(li=Mi,lMs).       % (dot)
modify([li=_ |lMs],li,N) ⟹ [li=N|lMs].                                 % (modify)
modify([li=Mi|lMs],l,N)  ⟹ [li=Mi|lMs_]    :- modify(lMs,l,N) ⟹ lMs_.% (modify)
case({[li=M]}:_, {lMs})  ⟹ (Mi $ M)        :- member(li=Mi,lMs).       % (case)
(M $ N)                  ⟹ (M_ $ N)        :- M ⟹ M_.
(M ! σ)                  ⟹ (M_ ! σ)        :- M ⟹ M_.
modify(M,l,N)            ⟹ modify(M_,l,N)  :- M ⟹ M_.
case(M, {lMs})           ⟹ case(M_, {lMs}) :- M ⟹ M_.

M ⟹* M_ :- M ⟹ M1,!, M1 ⟹* M_.
M ⟹* M.

% Free Type variables

ftv(x,[x]) :- x(x).
ftv(B,[]) :- b(B).
ftv((σ1->σ2),FTV) :- ftv(σ1,FTV1),ftv(σ2,FTV2),union(FTV1,FTV2,FTV).
ftv(lMs,FTVs) :- foldl([_=M,FTV,FTV_]>>(ftv(M,FTVi),union(FTV,FTVi,FTV_)),lMs,[],FTVs).
ftv({lMs},{FTVs}) :- foldl([_=M,FTV,FTV_]>>(ftv(M,FTVi),union(FTV,FTVi,FTV_)),lMs,[],FTVs).
ftv(∀(t,k,σ),FTV) :- kftv(k,KFTV),ftv(σ,QFTV),union(KFTV,QFTV,FTV1),subtract(FTV1,t,FTV).

kftv(u,[]).
kftv(lσs,FTVs) :- foldl([_:σ,FTV,FTV_]>>(ftv(σ,FTVi),union(FTV,FTVi,FTV_)),lσs,[],FTVs).
kftv({lσs},FTVs) :- foldl([_:σ,FTV,FTV_]>>(ftv(σ,FTVi),union(FTV,FTVi,FTV_)),lσs,[],FTVs).
tftv(lσs,FTVs) :- foldl([_:σ,FTV,FTV_]>>(ftv(σ,FTVi),union(FTV,FTVi,FTV_)),lσs,[],FTVs).

% Type system

(_,T) ▷ x : σ    :- member(x:σ,T).                            % VAR
(_,_) ▷ i : int  :- i(i).                                     % CONST
(_,_) ▷ true  : bool.                                         % CONST
(_,_) ▷ false : bool.                                         % CONST

(K,[x:σ1|T]) ▷ M1 : σ2
--%------------------------------------------------------------ ABS
(K,T) ▷ λ(x:σ1,M1) : (σ1->σ2).

(K,T) ▷ M1 : (σ1->σ2),  (K,T) ▷ M2 : σ1
--%------------------------------------------------------------ APP
(K,T) ▷ (M1 $ M2) : σ2.

tftv(T,FTV),  \+member(t,FTV),  ([t:k|K],T) ▷ M : σ
--%------------------------------------------------------------ TABS
(K,T) ▷ λ(t::k,M) : ∀(t,k,σ).

(K,T) ▷ M : ∀(t,k,σ1),  K ⊢ σ2:k,  tsub([σ2/t],σ1,σ1_)
--%------------------------------------------------------------ TAPP
(K,T) ▷ (M ! σ2) : σ1_.

maplist({K,T}/[li=Mi,li:σi]>>((K,T) ▷ Mi : σi), lMs, lσs)
--%------------------------------------------------------------ RECORD
(K,T) ▷ lMs : lσs.

(K,T) ▷ M : σ1,   K ⊢ σ1:[l:σ2]
--%------------------------------------------------------------ DOT
(K,T) ▷ (M # l) : σ2.

(K,T) ▷ M1 : σ1,  (K,T) ▷ M2 : σ2,  K ⊢ σ1:[l:σ2]
--%------------------------------------------------------------ MODIFY
(K,T) ▷ modify(M1,l,M2) : σ1.

(K,T) ▷ M : σ1,  K ⊢ σ2:{[l:σ1]}
--%------------------------------------------------------------ VARIANT
(K,T) ▷ ({[l=M]}:σ2) : σ2.

(K,T) ▷ M : {lσs},
maplist({K,T,σ}/[li=Mi,li:σi]>>((K,T) ▷ Mi : (σi->σ)),lMs,lσs)
--%------------------------------------------------------------ CASE
(K,T) ▷ case(M,{lMs}) : σ.

:- end_var_names(_).

File: scala/sos/sos2.scala
// Second-Order System Λ∀,#
// Syntaxs
object sos2 {
  type l = String
  type x = String
  trait k
  case object U extends k
  case class krecord(v:List[(l,σ)]) extends k
  case class kvariant(v:List[(l,σ)]) extends k
  trait σ
  case object tbool extends σ
  case object tint extends σ
  case class tx(v:String) extends σ
  case class tarr(v1:σ,v2:σ) extends σ
  case class trecord(v:List[(l,σ)]) extends σ
  case class tvariant(v:List[(l,σ)]) extends σ
  case class ∀(t:x,k:k,σ:σ) extends σ
  def t(t:σ) = t match {case tx(_)=>true case _=>false}
  def b(t:σ) = t == tbool || t == tint
  def q(a:Any) = a match {case _:σ => true case _ => false}
  def k(a:Any) = a match {case _:k => true case _ => false}
  trait M
  case object MTrue extends M
  case object MFalse extends M
  case class MInt(v:Int) extends M
  case class Mx(v:String) extends M
  case class Mλ(x:String,σ:σ,M:M) extends M
  case class MApp(M1:M,M2:M) extends M
  case class MTλ(t:String,k:k,M:M) extends M
  case class MTApp(M1:M,σ:σ) extends M
  case class MRecord(v:List[(l,M)]) extends M
  case class MDot(M:M,l:l) extends M
  case class MModify(M:M,l:l,M2:M) extends M
  case class MVariant(l:l,M:M,σ:σ) extends M
  case class MCase(M:M,w:List[(l,M)]) extends M
  
  def i(m:M) = m match {case MInt(_) => true case _ => false}
  def cb(m:M) = m match {case MTrue => true case MFalse => true case _ => i(m)}
  def x(m:M) = m match {case Mx(_)=>true case _=>false}
  def m(m:Any) = m match {case _:M=>true case _=>false}

  // Substitutions
  def tsub(S:Map[x,σ],σ:σ):σ = σ match {
    case tx(x) if S.contains(x) => tsub(S,S(x))
    case tx(x) => tx(x)
    case _ if b(σ) => σ
    case tarr(σ1,σ2) => tarr(tsub(S,σ1),tsub(S,σ1))
    case trecord(lMs) => trecord(lMs.map{case(l,t)=>(l,tsub(S,t))})
    case tvariant(lMs) => tvariant(lMs.map{case(l,t)=>(l,tsub(S,t))})
    case ∀(t,k,σ) => ∀(t,ksub(S-t,k),tsub(S-t,σ))
  }
  def ksub(S:Map[x,σ],k:k):k = k match {
    case U => U
    case krecord(lMs) => krecord(lMs.map{case(l,t)=>(l,tsub(S,t))})
    case kvariant(lMs) => kvariant(lMs.map{case(l,t)=>(l,tsub(S,t))})
  }
  def msub(S:Map[x,M],M:M):M = M match {
    case Mx(x) if S.contains(x) => msub(S,S(x))
    case Mx(x) => Mx(x)
    case m if cb(m) => m
    case Mλ(x,σ,m) => Mλ(x,σ,msub(S - x,m))
    case MApp(m1,m2) => MApp(msub(S,m1),msub(S,m2))
    case MTApp(m,σ) => MTApp(msub(S,m),σ)
    case MTλ(t,k,m) => MTλ(t,k,msub(S,m))
    case MRecord(lms) => MRecord(lms.map{case(l,m)=>(l,msub(S,m))})
    case MDot(m,l) => MDot(msub(S,m),l)
    case MModify(m1,l,m2) => MModify(msub(S,m1),l,msub(S,m2))
    case MVariant(l,m,σ) => MVariant(l,msub(S,m),σ)
    case MCase(m,lMs) => MCase(msub(S,m),lMs.map{case(l,m)=>(l,msub(S,m))})
  }
  def mtsub(S:Map[x,σ],M:M):M = M match {
    case Mλ(x,σ,m) => Mλ(x,tsub(S,σ),mtsub(S,m))
    case MApp(m1,m2) => MApp(mtsub(S,m1),mtsub(S,m2))
    case MTApp(m,σ) => MTApp(mtsub(S,m),tsub(S,σ))
    case MTλ(t,k,m) => MTλ(t,ksub(S-t,k),mtsub(S-t,m)) // todo to prolog
    case MRecord(lMs) => MRecord(lMs.map{case(l,m)=>(l,mtsub(S,m))})
    case MDot(m,l) => MDot(mtsub(S,m),l)
    case MModify(m1,l,m2) => MModify(mtsub(S,m1),l,mtsub(S,m2))
    case MVariant(l,m,σ) => MVariant(l,mtsub(S,m),tsub(S,σ))
    case MCase(m,lMs) => MCase(mtsub(S,m),lMs.map{case(l,m)=>(l,mtsub(S,m))})
    case _ => M
  }

  // Reduction rules
  def eval1(M:M):M = M match {
    case MApp(Mλ(x,_,m),n) => msub(Map(x->n),m)       // (β)
    case MTApp(MTλ(t,_,m),σ) => mtsub(Map(t->σ), m)   // (type-β)
    case MDot(MRecord(lMs),li) => lMs.toMap.apply(li) // (dot)
    case MModify(MRecord(lMs),l,m) => MRecord(lMs.map{case(li,mi)=>(li,if(l==li)m else mi)})// (modify)
    case MCase(MVariant(li,m,_),lMs) => MApp(lMs.toMap.apply(li),m) // (case)
    case MApp(m,n) => MApp(eval1(m),n)
    case MTApp(m,σ) => MTApp(eval1(m),σ)
    case MModify(m,l,n) => MModify(eval1(m),l,n)
    case MCase(m,lMs) => MCase(eval1(m),lMs)
  }
  def eval(m:M):M = try {
    eval(eval1(m))
  } catch {
    case _:Throwable => m
  }
  /*
  def kinding(K:Map[σ,σ],t:σ,k:k) {
    k match {
      case U =>
      case krecord(lσs) =>
        if(K.contains(t) && K(t) & k == k) else
        if()
      assert()
    }

    }
  }
  */
  /*
  // Kinding rules

  K ⊢     t : lσs  :- member(t:lσ2s,K), intersection(lσ2s,lσs,lσs).
  _ ⊢  lσ2s : lσs  :- intersection(lσ2s,lσs,lσs).
  K ⊢     t :{lσs} :- member(t:{lσ2s},K), intersection(lσ2s,lσs,lσs).
  _ ⊢ {lσ2s}:{lσs} :- intersection(lσ2s,lσs,lσs).
  _ ⊢     _ : u.
  */
  // Free Type variables
  def ftv(σ:σ):Set[x] = σ match {
    case tx(x) => Set(x)
    case _ if b(σ) => Set()
    case tarr(σ1,σ2) => ftv(σ1)++ftv(σ2)
    case trecord(lMs) => lMs.foldLeft(Set[x]()){case(tv,(_,m))=>tv++ftv(m)}
    case tvariant(lMs) => lMs.foldLeft(Set[x]()){case(tv,(_,m))=>tv++ftv(m)}
    case ∀(t,k,σ) => kftv(k)++ftv(σ) - t
  }
  def kftv(k:k):Set[x] = k match {
    case U => Set()
    case krecord(lMs) => lMs.foldLeft(Set[x]()){case(tv,(_,m))=>tv++ftv(m)}
    case kvariant(lMs) => lMs.foldLeft(Set[x]()){case(tv,(_,m))=>tv++ftv(m)}
  }
  def tftv(T:Map[x,σ]):Set[x] =
    T.foldLeft(Set[x]()){case(tv,(_,σ))=>tv++ftv(σ)}

  // Type system
  def ti(K:Map[x,k],T:Map[x,σ],M:M):σ = M match {
    case Mx(x)   => T(x)  // VAR
    case MInt(i) => tint  // CONST
    case MTrue   => tbool // CONST
    case MFalse  => tbool // CONST
    case Mλ(x,σ1,m1) => tarr(σ1,ti(K,T+(x->σ1),m1)) // ABS
    case MApp(m1,m2) => // APP
      ti(K,T,m1) match {
        case tarr(σ1,σ2) =>
          if (σ1 != ti(K,T,m2)) throw new Exception("error")
          σ2
      }
    case MTλ(t,k,m) => // TABS
      //val ftv = tftv(T)
      //if(!ftv.contains(t)) throw new Exception("error")
      val σ = ti(K+(t->k),T,m)
      ∀(t,k,σ)
    case MTApp(m,σ2) => // TAPP
      ti(K,T,m) match {
        case ∀(t,k,σ1) =>
          // kinding(K,σ2,k)
          tsub(Map(t->σ2),σ1)
      }
    case MRecord(lMs) => // RECORD
      trecord(lMs.map{case(li,mi)=>(li,ti(K,T,mi))})
    case MDot(m, l) => // DOT
      val σ1 = ti(K,T,m)
      //val k=kinding(K, σ1)
      //k.toMap.apply(l)
      σ1 match {
        case trecord(lts) => lts.toMap.apply(l)
      }
    case MModify(m1,l,m2) => // MODIFY
      val σ1 = ti(K,T,m1)
      val σ2 = ti(K,T,m2)
      //K ⊢ σ1:[l:σ2]
      σ1
    case MVariant(l,m,σ2) => // VARIANT
      val σ1=ti(K,T,m)
      // K ⊢ σ2:{[l:σ1]}
      σ2
    case MCase(m,lMs) => // CASE
      val t1 = ti(K,T,m)
      (t1,lMs.map{case(li,mi)=>(li,ti(K,T,mi))}) match {
        case (tvariant(vs),lts @ (_,tarr(_,σ))::_) =>
          lts.foreach {
            case(li,tarr(σi,σ_))=>
              if(vs.toMap.apply(li) != σi) throw new Exception("error")
              if(σ != σ_) throw new Exception("error")
            case (_,_) => throw new Exception("error")
          }
          σ
        case (_,_) => throw new Exception("error")
      }
  }
}
